File: apps/client-macos/Sources/TimeTrack/App/ManualIdleCoordinator.swift
import Foundation

/// The manual-session sibling of `AutoTrackingCoordinator`. Fed the same `WorkspaceObserver`
/// signals (via `FanOutSignalReceiver` in auto mode, or alone in manual mode), it drives a
/// `ManualIdleMonitor` and applies its one effect — the inactivity timeout — but ONLY while a
/// `.manual` session is live. All callbacks arrive on the main thread.
///
/// See `ManualIdleMonitor` for why a manual timer now stops on inactivity at all. In short: it
/// used to run through the away window and keep it unless the employee came back and said
/// otherwise, which is how a span reached 47 hours.
final class ManualIdleCoordinator: ManualIdleMonitorDelegate, AutoTrackingSignalReceiver {
    private let tracker: TimeTracker
    private let buffer: TimeEntryBuffering
    private let monitor: ManualIdleMonitor
    private let idGen: (Date) -> String
    /// Fired after the timeout has closed the live entry, so the owner can re-read `TimeTracker`.
    /// The menu bar indicator is driven by `MenuViewModel`, which cannot see a stop performed
    /// directly on the tracker — without this the icon would keep reporting a session that ended.
    /// Defaulted so the call sites and tests that don't care need no change.
    private let onTrackingStopped: () -> Void
    /// The manual entry the monitor is currently armed for. See `route`.
    private var armedEntryId: String?

    init(
        tracker: TimeTracker,
        buffer: TimeEntryBuffering,
        thresholdSeconds: Int,
        clock: @escaping () -> Date = Date.init,
        idGen: @escaping (Date) -> String = { UUIDv7.generate(now: $0) },
        onTrackingStopped: @escaping () -> Void = {}
    ) {
        self.tracker = tracker
        self.buffer = buffer
        self.monitor = ManualIdleMonitor(thresholdSeconds: thresholdSeconds, clock: clock)
        self.idGen = idGen
        self.onTrackingStopped = onTrackingStopped
        self.monitor.delegate = self
    }

    /// Sign-out / teardown. Nothing can be pending — a timeout disarms the monitor as it fires —
    /// so there is no half-resolved window to leak into the next session.
    func deactivate() { monitor.deactivate() }

    // MARK: AutoTrackingSignalReceiver (from WorkspaceObserver)

    func tick(idleSeconds: Int) { route { self.monitor.tick(idleSeconds: idleSeconds) } }
    func markAway() { route { self.monitor.markAway() } }
    /// Waking or unlocking is not itself a signal to do anything: if the session timed out while
    /// the machine slept, the entry is already closed and the person restarts when they are ready.
    func resume() {}

    /// Forward to the monitor only while a `.manual` session is live, arming it lazily on the
    /// first manual signal. This is also what re-arms after a timeout: the timeout disarms the
    /// monitor and closes the entry, so nothing routes again until the person starts a new manual
    /// session.
    ///
    /// Arming is per ENTRY, not merely "whenever the monitor is idle". The monitor stays armed
    /// through a Stop, so a Stop-then-Start would otherwise leave the NEW span measuring idleness
    /// from where the OLD one armed — and someone who stepped away for four minutes, came back,
    /// stopped and started a different project would watch that fresh entry close itself on the
    /// very next tick. That is exactly what `armedAt` exists to prevent, one level up.
    private func route(_ forward: () -> Void) {
        guard case .tracking(let entryId, _, _, .manual) = tracker.state else { return }
        if monitor.state == .inactive || armedEntryId != entryId {
            monitor.activate()
            armedEntryId = entryId
        }
        forward()
    }

    // MARK: ManualIdleMonitorDelegate

    func manualIdleMonitor(_ m: ManualIdleMonitor, didTimeOutFrom awayStart: Date, stoppingAt stopInstant: Date) {
        // Re-checked rather than assumed: the signal that triggered this was routed while a manual
        // session was live, but `TimeTracker` is the authority on what is running right now, and a
        // stop aimed at a session that already ended would close somebody else's span.
        guard case .tracking(_, _, _, .manual) = tracker.state else { return }
        tracker.stop(at: stopInstant)
        // The kept idle window, for the Idle panel. `.kept` because that is what the timesheet
        // says: these minutes are ON the entry. The stretch after `stopInstant` is untracked and
        // needs no record — there is no time to account for.
        if stopInstant > awayStart {
            IdleEventEnqueuer.enqueue(into: buffer, id: idGen(awayStart),
                                      from: awayStart, to: stopInstant, action: .kept)
        }
        onTrackingStopped()
    }
}

File: apps/client-windows/src/NiftyTimer/App/ManualIdleCoordinator.cs
using NiftyTimer.Storage;
using NiftyTimer.Tracking;

namespace NiftyTimer.App;

/// <summary>
/// The manual-session sibling of <see cref="AutoTrackingCoordinator"/>. Fed the same
/// <see cref="SessionObserver"/> signals — via <see cref="FanOutSignalReceiver"/> in auto mode, or
/// alone in manual mode — it drives a <see cref="ManualIdleMonitor"/> and applies its one effect,
/// the inactivity timeout, but ONLY while a MANUAL session is live. All callbacks arrive on the
/// UI thread.
///
/// See <see cref="ManualIdleMonitor"/> for why a manual timer now stops on inactivity at all. In
/// short: it used to run through the away window and keep it unless the employee came back and
/// said otherwise, which is how a span reached days long.
///
/// Note what this type no longer has: no away prompt, and no discard trim. Once the timeout has
/// closed the entry there is nothing left for the employee to adjudicate on return, so the prompt
/// — and with it the stale-prompt reconciliation, the entry-id snapshot and the display-clock
/// anchor that Discard needed — are gone by construction rather than by guard.
/// </summary>
public sealed class ManualIdleCoordinator : IManualIdleMonitorDelegate, ISignalReceiver
{
    private readonly TimeTracker _tracker;
    private readonly ITimeEntryBuffer _buffer;
    private readonly ManualIdleMonitor _monitor;
    private readonly Func<DateTimeOffset, string> _idGen;

    /// <summary>
    /// Fired after the timeout has closed the live entry, so the owner can re-read
    /// <see cref="TimeTracker"/>. The tray indicator is driven by <see cref="MenuViewModel"/>,
    /// which cannot see a stop performed directly on the tracker — without this the icon would
    /// keep reporting a session that policy already ended.
    /// </summary>
    private readonly Action _onTrackingStopped;

    public ManualIdleCoordinator(
        TimeTracker tracker,
        ITimeEntryBuffer buffer,
        int thresholdSeconds,
        Func<DateTimeOffset>? clock = null,
        Func<DateTimeOffset, string>? idGen = null,
        Action? onTrackingStopped = null)
    {
        _tracker = tracker;
        _buffer = buffer;
        _monitor = new ManualIdleMonitor(thresholdSeconds, clock);
        _idGen = idGen ?? (now => UuidV7.Generate(now));
        _onTrackingStopped = onTrackingStopped ?? (() => { });
        _monitor.Delegate = this;
    }

    public ManualIdleState MonitorState => _monitor.State;

    /// <summary>
    /// Sign-out / teardown. Nothing can be pending — a timeout disarms the monitor as it fires —
    /// so there is no half-resolved window to leak into the next session.
    /// </summary>
    public void Deactivate() => _monitor.Deactivate();

    public void Tick(int idleSeconds) => Route(() => _monitor.Tick(idleSeconds));

    public void MarkAway() => Route(_monitor.MarkAway);

    /// <summary>
    /// Waking or unlocking is not itself a signal to do anything: if the session timed out while
    /// the machine slept, the entry is already closed and the person restarts when they are ready.
    /// </summary>
    public void Resume()
    {
    }

    public void DidTimeOut(DateTimeOffset awayStart, DateTimeOffset stopInstant)
    {
        // Re-checked rather than assumed: the signal that triggered this was routed while a manual
        // session was live, but TimeTracker is the authority on what is running right now, and a
        // stop aimed at a session that already ended would close somebody else's span.
        if (_tracker.State is not TrackerState.Tracking { Source: TimeTracker.EntrySource.Manual })
        {
            return;
        }

        _tracker.Stop(stopInstant);

        // The kept idle window, for the Idle panel. KEPT because that is what the timesheet says:
        // these minutes are ON the entry. The stretch after stopInstant is untracked and needs no
        // record — there is no time to account for.
        if (stopInstant > awayStart)
        {
            IdleEventEnqueuer.Enqueue(_buffer, _idGen(awayStart), awayStart, stopInstant, ResolvedAction.Kept);
        }

        _onTrackingStopped();
    }

    private bool IsManualSessionLive =>
        _tracker.State is TrackerState.Tracking { Source: TimeTracker.EntrySource.Manual };

    /// <summary>
    /// Forward to the monitor only while a manual session is live, arming it lazily on the first
    /// manual signal. The guard is also what re-arms after a timeout: the timeout disarms the
    /// monitor and closes the entry, so nothing routes again until the person starts a new manual
    /// session.
    /// </summary>
    private void Route(Action forward)
    {
        if (!IsManualSessionLive)
        {
            return;
        }

        if (_monitor.State is ManualIdleState.Inactive)
        {
            _monitor.Activate();
        }

        forward();
    }
}
